git-server/src/lib.rs

```diff
@@ -24,19 +27,71 @@ pub struct Options {
     pub tls_cert: Option<PathBuf>,
     pub tls_key: Option<PathBuf>,
     pub git_receive_pack: bool,
+    pub authorized_keys: Vec<String>,
+    pub cert_nonce_seed: String,
+    pub allow_unauthorized_keys: bool,
+    pub hook_logs_dir: Option<String>,
+    pub log_hooks: bool,
 }
 
 #[derive(Debug, Clone)]
 pub struct Context {
     root: PathBuf,
     git_receive_pack: bool,
+    authorized_keys: Vec<String>,
+    cert_nonce_seed: String,
+    allow_unauthorized_keys: bool,
+    hook_logs_dir: Option<String>,
+    log_hooks: bool,
+}
+
+impl Context {
+    // Sets the config receive.advertisePushOptions, which lets the user known they can provide a push option `-o`,
+    // to specify unique attributes. This is currently not used, but may be used in the future.
+    pub fn advertise_push_options(&self) -> Result<(), Error> {
+        let field = "receive.advertisePushOptions";
+        let value = "true";
+
+        self.set_root_git_config(field, value)?;
+
+        Ok(())
+    }
+
+    // Enables users to submit a signed push: `push --signed`
+    pub fn set_cert_nonce_seed(&self) -> Result<(), Error> {
+        let field = "receive.certNonceSeed";
+        let value = &self.cert_nonce_seed;
+
+        self.set_root_git_config(field, value)?;
+
+        Ok(())
+    }
+
+    // updates the git config in the root project
+    pub fn set_root_git_config(&self, field: &str, value: &str) -> Result<(), Error> {
+        let mut path = self.root.clone();
+        path.push("git/config");
+
+        tracing::debug!("searching for git config at: {:?}", path);
+
+        let mut config = git2::Config::open(&path)?;
+
+        config.set_str(field, value)?;
+
+        Ok(())
+    }
 }
 
 /// Run the Git Server.
 pub async fn run(options: Options) {
     let ctx = Context {
         root: options.root,
```

Comment:
Just thinking through this more, not sure if makes sense that the git-server will convert a relative path to an absolute path. Where the git-server runs may not be in the `GIT_PROJECT_ROOT`, it could be run from anywhere.

I think it just makes more sense to pass an absolute path to the `--root` flag.

Reply:
I think there's some confusion here perhaps, what I'm suggesting is pretty simple: the user is responsible for passing a path to `--root`, whether relative or absolute that points to the right place. There's no need to force them to pass an absolute path. If they start `git-server` from `/foo/bar` and the monorepo is at `/foo/bar/monorepo`, then simply passing `monorepo` will do the trick.

Reply:
If we need to convert this to an absolute path for the hook, then simply using `Path::canonicalize` should be enough.

Reply:
> If we need to convert this to an absolute path for the hook, then simply using Path::canonicalize should be enough.

I wouldn't have thought it would be an issue for the hook, but after some tests, it appeared to be causing issues with a relative root path. dde24c9 uses `std::fs::canonicalize` for setting the `GIT_PROJECT_ROOT`, which resolves relative path problems when the git-http-backend executes the hooks.